### ISSUE TITLE
Detecting Unused Variable when being assign inside an IF statement

### ISSUE THREAD

**Issue by darthvader666uk** [sirbrillig/phpcs-variable-analysis#127]
Hi,

I was wondering if you could help?

Sometimes we have an IF statement where we also assign a Variable like this:

```
if (0 == ($user_id = get_current_user_id())) {
  // do stuff
}
```

This is correct for our code as we would then use the variable however, the sniff is detecting this as unused which (good coding practice or not) is technically correct.

Except for manually ignoring each time this is picked up, is there a setting which we can set in the XML to turn that section off?

Thanks for your time,

Stu

**Comment by sirbrillig** [comment]
Thanks for asking. This is a tricky question. If you search around for the topic of "assignment inside condition in php" you'll find this is often debated. The [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards) even have a rule against it for reasons that are outside the scope here.

I could argue that this small change is much more clear and less prone to errors:

```php
$user_id = get_current_user_id();
if (0 == $user_id ) {
  // do stuff
}
```

but you're right to point out that this sniff is not responsible for making such decisions on style, whatever I may think. All this should do is detect unused and undefined variables, so let's address that.

In your example, `$user_id` is in fact being used for comparison, but why does it exist at all? This version would have the same effect:

```php
if (0 == get_current_user_id()) {
  // do stuff
}
```

Maybe the example is lacking something. Assignments inside conditions have often been used as a shortcut to later using the variable in question, like this:

```php
if (0 == ($user_id = get_current_user_id())) {
  echo $user_id;
}
```

In that case, you shouldn't see any sort of warning from this sniff. (If I'm wrong and that's actually the problem you were reporting, please let me know.)

So: technically, yes, the expression `0 == ($user_id = get_current_user_id())` probably should be recognized as both assignment and use, but I can't think of any case where that would occur. I'm often wrong, though! If you could convince me otherwise, that would be very helpful.

**Comment by DavidAnderson684** [comment]
Sorry for wasting your time. We had indeed misunderstood the warning being generated. It was correctly identifying `$user_id` as *subsequently* unused, rather than incorrectly unidentifying it as previously undefined. One of you with the access to this issue should close it.

**Comment by sirbrillig** [comment]
👍 I'll close it, but if there's any more conversation you'd like to have, feel free to re-open.
